drivers/crypto/crypto_stm32.c

```diff
@@ -0,0 +1,400 @@
+	if (ret == 0) {
+		pkt->out_len = 16;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_cbc_encrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t vec[BLOCK_LEN_WORDS];
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)vec, sizeof(vec), iv, BLOCK_LEN_BYTES);
+	session->config.pInitVect = vec;
+
+	/* Prefix IV to ciphertext */
+	memcpy(pkt->out_buf, iv, 16);
+
+	ret = do_encrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf + 16);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len + 16;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_cbc_decrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t vec[BLOCK_LEN_WORDS];
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)vec, sizeof(vec), iv, BLOCK_LEN_BYTES);
+	session->config.pInitVect = vec;
+
+	ret = do_decrypt(ctx, pkt->in_buf + 16, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len - 16;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_ctr_encrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t ctr[BLOCK_LEN_WORDS] = {0};
+	int ivlen = ctx->keylen - (ctx->mode_params.ctr_info.ctr_len >> 3);
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)ctr, sizeof(ctr), iv, ivlen);
+	session->config.pInitVect = ctr;
+
+	ret = do_encrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_ctr_decrypt(struct cipher_ctx *ctx,
+				    struct cipher_pkt *pkt, u8_t *iv)
+{
+	int ret;
+	u32_t ctr[BLOCK_LEN_WORDS] = {0};
+	int ivlen = ctx->keylen - (ctx->mode_params.ctr_info.ctr_len >> 3);
+
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	copy_reverse_words((u8_t *)ctr, sizeof(ctr), iv, ivlen);
+	session->config.pInitVect = ctr;
+
+	ret = do_decrypt(ctx, pkt->in_buf, pkt->in_len, pkt->out_buf);
+	if (ret == 0) {
+		pkt->out_len = pkt->in_len;
+	}
+
+	return ret;
+}
+
+static int crypto_stm32_get_unused_session_index(void)
+{
+	int i;
+
+	for (i = 0; i < CRYPTO_MAX_SESSION; i++) {
+		if (!crypto_stm32_sessions[i].in_use) {
+			crypto_stm32_sessions[i].in_use = true;
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static int crypto_stm32_session_setup(struct device *dev,
+				      struct cipher_ctx *ctx,
+				      enum cipher_algo algo,
+				      enum cipher_mode mode,
+				      enum cipher_op op_type)
+{
+	int ctx_idx;
+	struct crypto_stm32_session *session;
+
+	if (ctx->flags & ~(CRYP_SUPPORT)) {
+		LOG_ERR("Unsupported flag");
+		return -EINVAL;
+	}
+
+	if (algo != CRYPTO_CIPHER_ALGO_AES) {
+		LOG_ERR("Unsupported algo");
+		return -EINVAL;
+	}
+
+	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
+	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
+	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
+		LOG_ERR("Unsupported mode");
```

Comment:
Please add a status clarifying CCM support.

Reply:
Thanks for your feedback.

Do you mean adding a comment that CCM is not supported, like

> Due to hardware restrictions CCM mode is not supported.

or a comment explaining why CCM is not supported, like

> The CRYP peripheral supports the AES ECB, CBC, CTR, CCM and GCM modes of operation, of which ECB, CBC, CTR and CCM are supported through the crypto API. However, in CCM mode, although the STM32Cube HAL driver follows the documentation (cf. RM0090, par. 23.3) by padding incomplete input data blocks in software prior encryption, incorrect authentication tags are returned for input data which is not a multiple of 128 bits. Therefore, CCM mode is not supported by this driver.

Reply:
Wait, @tbistm may have a hint to overcome the restriction.

Reply:
CCM can malfunction for several reasons. A wrong Tag may be due to :
1) a Tag with a variable length (not 16 bytes). It can be in theory (from NIST, A.1 Length Requirements) an element of {4, 6, 8, 10, 12, 14, 16} (bytes). but CRYP driven systematically reports its maximum size because it supposes memory is always reserved by upper layers.  
2) data alignment (including B0 and B1 should be 4 bytes align)

Reply:
So, the CRYP peripheral does not natively support the test vector used in the crypto sample, does it?
https://github.com/zephyrproject-rtos/zephyr/blob/1ab39e7c52237d5c438e11b59e2f58e00cba222e/samples/drivers/crypto/src/main.c#L366-L381
https://github.com/zephyrproject-rtos/zephyr/blob/1ab39e7c52237d5c438e11b59e2f58e00cba222e/samples/drivers/crypto/src/main.c#L392-L397
Is there any way to support partial inputs and tags with lengths other than 16 bytes?

There are other STM32 crypt drivers, I am not sure if they support other configurations:
- The [Linux driver](https://github.com/torvalds/linux/blob/0c99ee44b8921b872089a7374c733310d995fd92/drivers/crypto/stm32/stm32-cryp.c#L1158-L1175) at least looks like it can handle authentication tags smaller 16 bytes, but it targets the [STM32F756 and STM32MP1 SoCs](https://github.com/torvalds/linux/blob/0c99ee44b8921b872089a7374c733310d995fd92/drivers/crypto/stm32/stm32-cryp.c#L1907-L1909) only.
- The [wolfSSL driver](https://github.com/wolfSSL/wolfssl/blob/6b4551c012941ebc407d843c9a6dc835da546a34/wolfcrypt/src/aes.c#L5620-L5626) claims to generate the GCM "Auth correctly for partial or when IV != 12". Of course, GCM authentication tag generation differs from the one for CCM.

Reply:
I forgot the STM32/X-CUBE Crypto Library. It is [NIST CAVP-certified](https://www.st.com/content/ccc/resource/sales_and_marketing/presentation/product_presentation/group0/79/b4/63/20/21/53/47/54/fips_cavp_certification/files/fips_cavp_certification.pdf/jcr:content/translations/en.fips_cavp_certification.pdf), uses the CRYP peripheral for AES (cf. [table 4 on page 26 in UM1924](https://www.st.com/content/ccc/resource/technical/document/user_manual/group0/f9/6e/f2/a2/b4/ec/49/c0/DM00215061/files/DM00215061.pdf/jcr:content/translations/en.DM00215061.pdf#page=26)) on the mentioned STM32F4 devices, supports nonces with sizes from 7 to 13 bytes, authentication tags with sizes from 4 to 16 bytes, and should work on complete and partial input data.

Reply:
I suggest to limit the SW without CCM. Your comment is relevant.
There might be a problem and we have been investigating.

Reply:
> Please add a status clarifying CCM support.

Comment added.

Reply:
Sorry fro late feedback, can you pout your later proposition instead? (the detailed one)
I'd like to avoid that someone reads this and thinks this was a mistake after a quick check in the ref manual.

Reply:
Comment has been changed to the detailed one.

Reply:
> There might be a problem and we have been investigating.

Thanks. Please keep me posted.